lib/ui/login/login_view_model.dart

```diff
@@ -0,0 +1,33 @@
+import 'package:flutter_survey/model/login_model.dart';
+import 'package:flutter_survey/usecases/base/base_use_case.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_survey/ui/login/login_state.dart';
+import 'package:flutter_survey/usecases/auth_token_storage_use_case.dart';
+import 'package:flutter_survey/usecases/log_in_use_case.dart';
+
+class LoginViewModel extends StateNotifier<LoginState> {
+  final LogInUseCase _loginUseCase;
+  final AuthTokenStorageUseCase _authTokenStorageUseCase;
+
+  LoginViewModel(
+    this._loginUseCase,
+    this._authTokenStorageUseCase,
+  ) : super(const LoginState.init());
+
+  void logIn(String email, String password) async {
+    state = const LoginState.loading();
+    final LoginInput input = LoginInput(
+      email: email,
+      password: password,
+    );
+    final result = await _loginUseCase.call(input);
+    if (result is Success<LoginModel>) {
+      final LoginModel loginData = result.value;
+      _authTokenStorageUseCase.save(loginData);
+      state = const LoginState.loginSuccess();
```

Comment:
What happens if `_authTokenStorageUseCase.save` fails to save the data, please? 

I suggested you guys move the data-saving logic to LogInUseCase because of this issue. We can just catch the final error for both actions. By separating the actions into 2 separate use cases, we need to catch and cover both cases here 🙏

Reply:
@luongvo @Tuubz please share your opinion
- do you have any idea that how do we know if data saving operation in secure storage is failed or not? the `write` method of `secure_storage` doesn't throw error or exception. That's why I just write test to check if the method get invoked or not. 
- For refresh token task we will receive same response as login response and we need to save from same data like `acessToken`, `refreshTken`, `tokenType`. As in refresh token task we don't use `LogInUseCase` we have to write same data saving code in `RefreshTokenUseCase` if we don't have  separate use case for data saving, that's why I think separate use case for data saving can reduce redundant code.

Reply:
> do you have any idea that how do we know if data saving operation in secure storage is failed or not? 

I think we could wrap it with a `try/catch`:

```
 try {
     await _secureStorage.writeSecureData(accessTokenKey, data.accessToken);
     await _secureStorage.writeSecureData(refreshTokenKey, data.refreshToken);
     await _secureStorage.writeSecureData(tokenTypeKey, data.tokenType);
    return Success(oauthLogin);
  } catch (exception) {
    return Failed(UseCaseException(exception));
  }
```

>  why I think separate use case for data saving can reduce redundant code.

I agree with this too, and as @Shayokh144 mentioned before, use-cases should have only one responsibility.

Reply:
So I would suggest:

1. Keep it in two separate use-cases.
2. Wrap `try/catch` 
3. Print out a message , if storing access token failed (it shouldn't block the user from using the app)

Reply:
Oh.. wait 🙈 

The user will be blocked anyway, so we should show an error snack bar instead (my bad)

Reply:
@Shayokh144 
- Your first question is answered by https://github.com/nimblehq/ic-flutter-taher-toby/pull/70#discussion_r1172058791.
- Your second concern: Yes, it's not wrong to have 2 separate use cases. Then, please do testing properly with the chosen way 🙏

Reply:
@Shayokh144 I think you can skip this review as `_authTokenStorageUseCase` returns void, not a Result.